### PR TITLE
Add clang tooling

### PR DESCRIPTION
**Pull request by NewProggie**
- Normalize the license boilerplate, redirecting to the already existing LICENSE file in the root of the project. This reduces clutter and is common practice (as seen in the Chromium project, Mozilla and others)
- Add clang-format configuration file based on the Google coding style guide, using the most current version 3.9
- Apply all reformatting based on the coding style guide

### CODE REVIEW

File: tiny_dnn/core/backend.h [hunk 1]
```diff
@@ -40,118 +18,128 @@ class context;
...
-    switch (type) {
-        case backend_t::tiny_dnn: os << "TinyDNN"; break;
-        case backend_t::nnpack:   os << "NNPACK";  break;
-        case backend_t::libdnn:   os << "LibDNN";  break;
-        case backend_t::avx:      os << "AVX";     break;
-        case backend_t::opencl:   os << "OpenCL";  break;
-        default:
-            throw nn_error("Not supported ostream enum.");
-            break;
-    }
```
Comment (edgarriba) [inline tiny_dnn/core/backend.h]: @NewProggie can we keep also this format here?

File: tiny_dnn/core/backend.h [hunk 2]
```diff
@@ -40,118 +18,128 @@ class context;
...
+  // context holds solution-dependent parameters
+  // context should be able to hold any types of structures (like boost::any)
+  explicit backend(context* ctx_ = nullptr) {}
 
-    // core math functions
+  // core math functions
 
-    virtual void conv2d(const std::vector<tensor_t*>& in_data,
-                        std::vector<tensor_t*>&       out_data) = 0;
+  virtual void conv2d(const std::vector<tensor_t*>& in_data,
```
Comment (edgarriba) [inline tiny_dnn/core/backend.h]: @NewProggie and here parameter alignment?

**Comment by edgarriba** [comment]
@NewProggie in the first error I see that it complains about that `layerptr_t` doesn't exist 
check that you didn't remove this: https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/node.h#L54

File: test/test_average_pooling_layer.h [hunk 3]
```diff
@@ -1,111 +1,73 @@
...
-        0, 1, 2, 3,
-        8, 7, 5, 6,
-        4, 3, 1, 2,
-        0,-1,-2,-3
-    };
-
-    vec_t expected = {
-        4, 4,
-        1.5, -0.5
-    };
```
Comment (NewProggie) [inline test/test_average_pooling_layer.h]: As I said, we should probably enwrap this with
```
// clang-format off
...
// clang-format on
```

File: tiny_dnn/core/backend.h (hunk 1 above)
Comment (NewProggie) [inline tiny_dnn/core/backend.h]: What do you mean exactly? Leaving the `break` statement on the same line? In this case, see my comment above.

File: test/test_average_pooling_layer.h (hunk 3 above)
Comment (edgarriba) [inline test/test_average_pooling_layer.h]: Oh, I see 😃

**Comment by edgarriba** [comment]
@NewProggie Please rebase from master

**Comment by edgarriba** [comment]
@NewProggie I've found this useful link to integrate clang-format and clang-tidy with Cmake
http://www.labri.fr/perso/fleury/posts/programming/using-clang-tidy-and-clang-format.html

**Comment by edgarriba** [comment]
@NewProggie Any update to this PR?

**Comment by bhack** [comment]
@NewProggie we want to try to merge this to master before opening a new branch that will touch many files. Can you give us a feedback?

**Comment by bhack** [comment]
@Randl still you need to finalize at https://github.com/tiny-dnn/tiny-dnn/pull/368 or you can converge the efforts here?

**Comment by Randl** [comment]
@bhack  Of course automated code style sounds much better

**Comment by bhack** [comment]
@randl Ok we are going to integrating quite soon the new Tensors with a series of PRs and I think that if we start to merge that we will charge this PR of an heavy rebasing effort. So let me know how this could be handled without wasting to much efforts.

**Comment by pansk** [comment]
Other conflicting style changes have been applied to the master already, I think we should close this PR.

**Comment by Randl** [comment]
@pansk Yeah, but since it's automatic, simple rebasing and re-running would make it good to merge. `Stalled` label in hope someone will pick it is OK

**Comment by bhack** [comment]
Yes the main advantage of this is that we could re-running.

File: test/test_average_pooling_layer.h (hunk 3 above)
Comment (beru) [inline test/test_average_pooling_layer.h]: Writing clang-format option comments in source code sounds a bit weird.
Do we have clear formatting rules before applying `clang-format`?

**Comment by edgarriba** [comment]
I think that after https://github.com/tiny-dnn/tiny-dnn/pull/513, we can close this PR

**Comment by Randl** [comment]
Agreed

**Pull request closed without merging**